### ISSUE TITLE
How can I get more documentation on Scenes, Stages and stuff?

### ISSUE THREAD

**Issue by takdw** [telegraf/telegraf#705]
Hi all, I was checking out the docs and noticed that it doesn't say a lot about Scenes and Stages. Where can I get more detailed docs since I'm new to this module?

**Comment by dvgololobov** [comment]
You can use scenes like some microbot instance. [Examples](https://github.com/telegraf/telegraf/tree/master/docs/examples).

**Comment by IMalyugin** [comment]
I ran into bagful of troubles implementing Scenes myself, the main problem is lack of working examples or even detailed issues regarding the question. To help other people having similar issues, I'll just leave this tutorial here.

**What are Scenes and SceneWizard, what are their purpose?**
Basic API of telegraf allows for implementation of simple interactions, such as:
command -> response or message -> response, trigger -> response

While that is quite powerful and easy to use, there is a set of tasks, that require to create a DIALOGUE between user and the bot. That is where Scenes and SceneWizards fill in.


**What is a Scene, how does it work?**
Scene is like a namespace, an abstract isolated room that user can be pushed into via code, when inside one, all the outside noise such as global commands or message handlers stop reacting on user actions.
Scenes can be entered, or left from, scenes can have their own command/message listeners, that only work while user is inside.


**What kind of Scenes are there, what's the difference?**
There are two types of Scenes, a more configurable `BaseScene`, or a less configurable, Step-based `SceneWizard`.


**How to create and use a `BaseScene`**
`BaseScene` is just an isolated namespace that needs manually attached hooks, here is an example:
```
import { BaseScene } from 'telegraf';

const scenarioTypeScene = new BaseScene('SCENARIO_TYPE_SCENE_ID');

scenarioTypeScene.enter((ctx) => {
  ctx.session.myData = {};
  ctx.reply('What is your drug?', Markup.inlineKeyboard([
    Markup.callbackButton('Movie', MOVIE_ACTION),
    Markup.callbackButton('Theater', THEATER_ACTION),
  ]).extra());
});

scenarioTypeScene.action(THEATER_ACTION, (ctx) => {
  ctx.reply('You choose theater');
  ctx.session.myData.preferenceType = 'Theater';
  return ctx.scene.enter('SOME_OTHER_SCENE_ID'); // switch to some other scene
});

scenarioTypeScene.action(MOVIE_ACTION, (ctx) => {
  ctx.reply('You choose movie, your loss');
  ctx.session.myData.preferenceType = 'Movie';
  return ctx.scene.leave(); // exit global namespace
});

scenarioTypeScene.leave((ctx) => {
  ctx.reply('Thank you for your time!');
});

// What to do if user entered a raw message or picked some other option?
scenarioTypeScene.use((ctx) => ctx.replyWithMarkdown('Please choose either Movie or Theater'));
```

**How to create and use a WizardScene**
`SceneWizard` is a lot easier to use, it does not require any hooks, although it might have them.
SceneWizard is created in via the following call
`const wizardScene = new WizardScene(wizardSceneId, ...stepHandlers);`

Context (ctx) inside wizardScene is enhanced with `ctx.wizard` field. That is step control interface, it has following fields:
ctx.wizard.next() - to advance to next step;
ctx.wizard.prev() - to go back to previous step;
ctx.wizard.cursor - to get current step index;
ctx.wizard.selectStep(index) - to jump straight to a given step index, can be used to implement branching;

Normally each stepHandler, should end with ctx.wizard.next(); Once called, bot awaits for the next trigger from user, such as message or action, the result will be available in the next stepHandler.

If there was an error or ctx.wizard.next() was not called, current step is reentered and bot awaits for the next trigger from user. This can be used to implement validation

Example:
```
import WizardScene from 'telegraf/scenes/wizard';

const contactDataWizard = new WizardScene(
  'CONTACT_DATA_WIZARD_SCENE_ID', // first argument is Scene_ID, same as for BaseScene
  (ctx) => {
    ctx.reply('What is your name?');
    ctx.scene.session.contactData = {};
    return ctx.wizard.next();
  },
  (ctx) => {
    // validation example
    if (ctx.message.text.length < 2) {
      ctx.reply('Please enter name for real');
      return; 
    }
    ctx.scene.session.contactData.fio = ctx.message.text;
    ctx.reply('Enter your e-mail');
    return ctx.wizard.next();
  },
  async (ctx) => {
    ctx.scene.session.contactData.email = ctx.message.text;
    ctx.reply('Thank you for your replies, we'll contact your soon');
    await mySendContactDataMomentBeforeErase(ctx.scene.session.contactData);
    return ctx.scene.leave();
  },
);


**Is there a way to gain more control over stepHandlers**
In fact there is. A step handler may be created as a Composer, this way you can attach multiple hooks to it, to have different response, based on the data.


**How are Scenes attached to telegraf bot?**
The founding instance of Scenes is a `Stage`, 
```
import { Stage } from 'telegraf';
const stage = new Stage([scene1, scene2, scene3, ...]);
bot.use(session()); // to  be precise, session is not a must have for Scenes to work, but it sure is lonely without one
bot.use(stage.middleware());
```

Stage itself is quite useless, or even harmful, because a lot of it's API is malfunctioning, but you may think of it as a boilerplate to concatenate scenes into a middleware.


**How to enter a scene (BaseScene or WizardScene) **
There are countless examples of how to enter scene, but all of them do not work, lets list some of them:
1.  NOT LIKE THIS: `Stage.enter('CONTACT_DATA_WIZARD_SCENE_ID');` - this piece of code simply doesn't work. Why? Maybe because something broke down in telegraf recently.
2. NOT LIKE THIS EITHER: ` const stage = new Stage([scene1, scene2, ...], { default: 'CONTACT_DATA_WIZARD_SCENE_ID' });` - this is a quite nasty example, provided in official documentation. When default key is provided as option creating a Stage, user automatically enters the given scene, but there is no way of leaving scenes anymore. As soon as you call `ctx.scene.leave()`, you end up in the default scene.

The simple and reliable way of entering scenes is the same as switching between different scenes.
`ctx.scene.enter('CONTACT_DATA_WIZARD_SCENE_ID')` - this code is not limited to the scope of current scene, it's available anywhere, once you enable stage.middleware(), global context is extended with `ctx.scene`, so you can freely use it anywhere in your application.


**How to persist data between Scenes**
If your bot has been extended with session middleware, you can always read/write data via `ctx.session` field. If you are using WizardScene there is yet another option - `ctx.wizard.session` - it is similar except, once user leaves the WizardScene steps, it gets erased. So it is not an option if you combine BaseScenes and WizardScenes in a single scenario;


*** Final words ***
Dear Reader, I wrote all the stuff above inventing code on the fly, it has not been tested so small bugs might crawl in when copy pasted, and for that I am sorry. But all the info is actually well tested and will help you avoid the hole in documentation i've been struggling with, while digging in the underlying code.

**Comment by IMalyugin** [comment]
I ran into bagful of troubles implementing Scenes myself, the main problem is lack of working examples or even detailed issues regarding the question. To help other people having similar issues, I'll just leave this tutorial here.

**What are Scenes and WizardScene, what are their purpose?**
Basic API of telegraf allows for implementation of simple interactions, such as:
command -> response or message -> response, trigger -> response

While that is quite powerful and easy to use, there is a set of tasks, that require to create a DIALOGUE between user and the bot. That is where Scenes and WizardScenes fill in.


**What is a Scene, how does it work?**
Scene is like a namespace, an abstract isolated room that user can be pushed into by means of code. When inside one, all the outer noise,  such as global commands or message handlers, stop reacting on user actions.
Scenes can be entered, or left from, scenes can have their own hooks (command/message listeners), that only work while user is inside.


**What kind of Scenes are there, what's the difference?**
There are two types of Scenes, a more configurable `BaseScene`, or a less configurable, Step-based `WizardScene`.


**How to create and use a BaseScene**
`BaseScene` is just an isolated namespace that needs manually attached hooks, here is an example:
```ts
import { Scenes } from 'telegraf';

const scenarioTypeScene = new Scenes.BaseScene('SCENARIO_TYPE_SCENE_ID');

scenarioTypeScene.enter((ctx) => {
  ctx.session.myData = {};
  ctx.reply('What is your drug?', Markup.inlineKeyboard([
    Markup.callbackButton('Movie', MOVIE_ACTION),
    Markup.callbackButton('Theater', THEATER_ACTION),
  ]).extra());
});

scenarioTypeScene.action(THEATER_ACTION, (ctx) => {
  ctx.reply('You choose theater');
  ctx.session.myData.preferenceType = 'Theater';
  return ctx.scene.enter('SOME_OTHER_SCENE_ID'); // switch to some other scene
});

scenarioTypeScene.action(MOVIE_ACTION, (ctx) => {
  ctx.reply('You choose movie, your loss');
  ctx.session.myData.preferenceType = 'Movie';
  return ctx.scene.leave(); // exit global namespace
});

scenarioTypeScene.leave((ctx) => {
  ctx.reply('Thank you for your time!');
});

// What to do if user entered a raw message or picked some other option?
scenarioTypeScene.use((ctx) => ctx.replyWithMarkdown('Please choose either Movie or Theater'));
```

**How to create and use a WizardScene**
`WizardScene` is a lot easier to use, it does not require any hooks, although it might have them.
WizardScene is created in via the following call
`const wizardScene = new WizardScene(wizardSceneId, ...stepHandlers);`

Context (`ctx`) inside wizardScene is enhanced with `ctx.wizard` field. That is step control interface, it has following fields:
`ctx.wizard.next()` - to advance to next step;
`ctx.wizard.back()` - to go back to previous step;
`ctx.wizard.cursor` - to get current step index;
`ctx.wizard.selectStep(index)` - to jump straight to a given step index, can be used to implement branching;

Normally each stepHandler, should end with `ctx.wizard.next();` Once called, bot awaits for the next trigger from user, such as message or action, the result will be available in the next stepHandler.

If there was an error or `ctx.wizard.next()` was not called, current step is reentered and bot awaits for the next trigger from user. This can be used to implement validation

Example:
```ts
import { Scenes } from 'telegraf';

const contactDataWizard = new Scenes.WizardScene(
  'CONTACT_DATA_WIZARD_SCENE_ID', // first argument is Scene_ID, same as for BaseScene
  (ctx) => {
    ctx.reply('What is your name?');
    ctx.wizard.state.contactData = {};
    return ctx.wizard.next();
  },
  (ctx) => {
    // validation example
    if (ctx.message.text.length < 2) {
      ctx.reply('Please enter name for real');
      return; 
    }
    ctx.wizard.state.contactData.fio = ctx.message.text;
    ctx.reply('Enter your e-mail');
    return ctx.wizard.next();
  },
  async (ctx) => {
    ctx.wizard.state.contactData.email = ctx.message.text;
    ctx.reply('Thank you for your replies, we'll contact your soon');
    await mySendContactDataMomentBeforeErase(ctx.wizard.state.contactData);
    return ctx.scene.leave();
  },
);
```

**Is there a way to gain more control over stepHandlers**
In fact there is. A step handler may be created as a Composer, this way you can attach multiple hooks to it, to have different response, based on the data.


**How are Scenes attached to telegraf bot?**
The foundational component of Scenes is a `Stage`, 
```ts
import { Scenes } from 'telegraf';

const stage = new Scenes.Stage([scene1, scene2, scene3, ...]);
bot.use(session()); // to  be precise, session is not a must have for Scenes to work, but it sure is lonely without one
bot.use(stage.middleware());
```

Stage itself is just there for the code sugar, you may think of it as a boilerplate to concatenate scenes into a middleware.


**How to enter a scene (BaseScene or WizardScene)**
Entering a scene can be a little tricky if you do not pay attention, the following code can be found in the original example:
1. NOT RECOMMENDED:
```ts
const stage = new Stage([scene1, scene2, ...], { default: 'CONTACT_DATA_WIZARD_SCENE_ID' });
```
this is a quite nasty way of entering scene. When default sceneId is provided as an option to Stage constructor, user automatically enters provided scene and there is no way of leaving scenes anymore. As soon as you call `ctx.scene.leave()`, you end up in the default scene.

2. CAUTION ADVISED:

```ts
Stage.enter('CONTACT_DATA_WIZARD_SCENE_ID');
```
this piece of code looks good and it is valid, but when you try using it, it doesn't work. Why? Because you're probably using it as a function, but it is in fact a middleware factory! So instead of doing this:
```ts
// BAD
bot.hears('hi', () => {
  Stage.enter('CONTACT_DATA_WIZARD_SCENE_ID');
});
```
```ts
// ALSO BAD
bot.hears('hi', (ctx) => Stage.enter('CONTACT_DATA_WIZARD_SCENE_ID'));
```
You have to do this:
```ts
// GOOD
bot.hears('hi', Stage.enter('CONTACT_DATA_WIZARD_SCENE_ID'));
```

3. Recommended
```ts
ctx.scene.enter('CONTACT_DATA_WIZARD_SCENE_ID')
```
is the simple and reliable way of entering scenes. This code is not limited to the scope of current scene, it's available anywhere, once you enable stage.middleware(), global context is extended with `ctx.scene`, so you can freely use it anywhere in your application.

The same API (as 2 and 3) can be used to switch from one scene to another.


**How to persist data between Scenes**
If your bot has been extended with session middleware, you can always read/write data via `ctx.session` field. If you choose this path, you are responsible for initializing and cleaning up your session data.

If you are using WizardScene there is another option:
```ts
ctx.wizard.state
// or 
ctx.scene.state
```
Actually these fields are almost identical, as they point to the same data object:
`ctx.wizard.state` -> `ctx.scene.state` -> `ctx.session.state`
The only difference is - first two automatically resolve to {}.

If you follow this path, you don't have to clean the data, as it is cleaned up automatically once you leave a scene or enter a new one.

If you combine BaseScenes with WizardScenes, state will be erased upon switching from one scene to another, but there is still a way:
You can pass state to another scene via second argument to `ctx.scene.enter(sceneId, initialState)`. So if you use `ctx.state`, you can call `ctx.scene.enter('MY_SCENE_ID', ctx.scene.state)`, and your state will be written into state of new scene.


**To sum up**
Dear Reader, I wrote all the stuff above inventing code on the fly, it has not been tested so small bugs might crawl in when copy pasted, and for that I am sorry. But all the info is actually well tested and will help you avoid the hole in documentation i've been struggling with, while digging in the underlying code.

**Comment by dotcypress** [comment]
@takdw I love your explanation, so clear and easy to understand. Will be great to add this to [help page](https://github.com/telegraf/telegraf/blob/develop/docs/README.md). 

One thing about `Stage.enter('CONTACT_DATA_WIZARD_SCENE_ID')`, this functions is a middleware factory, and used only for code sugaring. 
Following two examples are logically identical:

```js
greeterScene.hears('hi', Stage.enter('greeter'))

greeterScene.hears('hi', (ctx) => ctx.stage.enter('greeter'))
```

**Comment by IMalyugin** [comment]
@dotcypress thanks, hope it'll help people find what they're looking for faster :)

As for the Stage.enter('...'), I tried using it from Composer, then straight from original bot instance with no avail, for some reason it just didn't enter the scene no matter what. I even tried ctx.stage.enter() - with the same result. Only ctx.scene.enter() worked for me, that's why I called Stage harmful, after all the first thing that comes to mind when you want to enter a scene from outside - is that you have to use the main door, a scene aggregator - Stage.

Put simple, my app looks like this:
```
const bot = new Telegraf(BOT_TOKEN);
bot.telegram.setWebhook(`${BOT_URL}/bot${BOT_TOKEN}`);
bot.startWebhook(`/bot${BOT_TOKEN}`, 
bot.use(session());
bot.use(reportWizard.middleware());

-- reportWizard.js
...
const reportWizard = new WizardScene(REPORT_WIZARD, ...);

const bot = new Composer();

const stage = new Stage([reportWizard]);
bot.use(stage.middleware());

// does not work
// bot.command('wizard', (ctx) => ctx.stage.enter(REPORT_WIZARD));

// does not work
// bot.command('wizard', (ctx) => Stage.enter(REPORT_WIZARD));

// works
bot.command('wizard', (ctx) => ctx.scene.enter(REPORT_WIZARD));

export default bot;
```

Maybe Stage.enter doesn't work from within another Composer, or there's an issue somewhere.

**Comment by IMalyugin** [comment]
I would like to share my enhancement of Scene mechanics, that allows for better code due to higher responsibility separation.

What I was trying to solve is find a way to **properly implement scene branching**.

I have a complicated dialogue between user and the bot, that depends on already available data and on the steps user takes, so it is not a linear wizard with 10 steps out of 10. Things get even more complicated as there are 2 scenarios, with mostly same steps.

What is wrong with current Scene mechanics, that make it complicated to implement?
Here's a list of things, that block out advance:
1. To implement a complicated scenario, we need to be able to define steps separately, and then join them together with a higher abstraction, but each step mixes two different responsibilities: validating n' accepting previous step data and showing reply or introduction for the next step;
2. Switching between steps is only possible via step index. But stepIndex is not a constant value, it's not safe to use, it will change if we add another scene; If we decide to use each step as BaseScene - there's just too much code to be written;
3. What if we want to put in two similar scenes? Scenes are linked with sceneId, so if we put two of the same scenes, it will not work.
4. How do we await for scene to end, so we can run code afterwards?

How do we overcome all these issues? With a factory of course.
Here's the code for SceneFactory. It's not perfect and does not claim to be a library.
```
import WizardScene from 'telegraf/scenes/wizard';

const unwrapCallback = async (ctx, nextScene) => {
  const nextSceneId = await Promise.resolve(nextScene(ctx));
  if (nextSceneId) return ctx.scene.enter(nextSceneId, ctx.scene.state);
  return ctx.scene.leave();
};

/**
 * Takes steps as arguments and returns a sceneFactory
 *
 * Additionally does the following things:
 * 1. Makes sure next step only triggers on `message` or `callbackQuery`
 * 2. Passes second argument - doneCallback to each step to be called when scene is finished
 */
export const composeWizardScene = (...advancedSteps) => (
  /**
   * Branching extension enabled sceneFactory
   * @param sceneType {string}
   * @param nextScene {function} - async func that returns nextSceneType
   */
  function createWizardScene(sceneType, nextScene) {
    return new WizardScene(
      sceneType,
      ...advancedSteps.map((stepFn) => async (ctx) => {
        /** ignore user action if it is neither message, nor callbackQuery */
        if (!ctx.message && !ctx.callbackQuery) return undefined;
        return stepFn(ctx, () => unwrapCallback(ctx, nextScene));
      }),
    );
  }
);
```

**How does it work, what does it do?**
Lets look at an example:
```
// scenes/contactData.js
import { composeWizardScene } from '../../utils/sceneFactory';

export const createContactDataScene = composeWizardScene(
  (ctx) => {
    ctx.reply('Please enter your credentials');
    return ctx.wizard.next();
  },
  (ctx, done) => {
    ctx.wizard.state.credentials = ctx.message.text;
    return done();
  },
);
```
In the above example, we created a sceneFactory, using nearly same interface as WizardScene, what are the difference:
1. We did not define a sceneId (or sceneType as I call it);
2. There is a new `done` function-argument in call to each step;

Not impressed? Wait till we see how it works on the stage;
```
// reportStageWizard.js
import { createEntryScene } from '../scenes/entryData';
import { createContactDataScene } from '../scenes/contactData';
import { createOrderDataScene } from '../scenes/orderData';
import { createActionTimeScene } from '../scenes/actionTimeScene';
import { ENTRY_SCENE, CONTACT_DATA_SCENE, ORDER_DATA_SCENE, ACTION_TIME_SCENE } from '../scenes/sceneTypes';

const bot = new Composer();

const stage = new Stage([
  createEntryScene(ENTRY_SCENE, (ctx) => (ctx.session.hasCreds ? ORDER_DATA_SCENE : CONTACT_DATA_SCENE)),
  createContactDataScene(CONTACT_DATA_SCENE, () => ORDER_DATA_SCENE),
  createOrderDataScene(ORDER_DATA_SCENE, () => ACTION_TIME_SCENE),
  createActionTimeScene(ACTION_TIME_SCENE, async (ctx) => {
    const { fio, date, time, clientName, actionType } = ctx.wizard.state;
    await sendReport({ date, time, clientName, actionType, fio, payment: '?' });
    await ctx.reply('Thank you for your report');
  }),
]);

bot.use(stage.middleware());
bot.command('test', Stage.enter(ENTRY_SCENE));
```

Let me first explain how `create***Scene` works. It takes two arguments:
`sceneType` - is sceneId - a string identifier for the scene;
`nextScene` - is the wrapped `done` callback function that we call once scene finishes, current ctx is forced in its call. nextScene is used to control the flow of our stage, as it awaits for new sceneId to be returned. If that happens current ctx.wizard.state is automatically passed to ctx.scene.enter call. If it returns nullable value, then ctx.scene.leave() is called.

Now our stage actually has a meaning, it joins multiple scenes into a same large scenario. And now we can split each scene logic into a separate file. Scene naming and branching is done in a single place, it can be both simple or complicated.

As an additional feature, I've added verification into each step - so that step function is only fired if there actually was a message or callbackButton click from the user.


I've been trying to figure out where to post this, but I guess it's best placed near the original explanation above.

**Comment by takdw** [comment]
@IMalyugin Thank you very much for taking the time and explaining it in depth. I sure hope some of it gets added to the docs. Thank you again.

**Comment by 1valdis** [comment]
As seen by the amount of upvotes, the demand for clearer docs on Scenes, Stages and stuff is quite high. No progress on that yet, beside @IMalyugin's terrific guide? I'd appreciate the mentioned guide or something similar to be directly in the docs.

**Comment by foodornt** [comment]
> Context (`ctx`) inside wizardScene is enhanced with `ctx.wizard` field. That is step control interface, it has following fields:
> `ctx.wizard.next()` - to advance to next step;
> `ctx.wizard.prev()` - to go back to previous step;
> `ctx.wizard.cursor` - to get current step index;
> `ctx.wizard.selectStep(index)` - to jump straight to a given step index, can be used to implement branching;
little correction, `ctx.wizard.back()`, not .prev()
and thank you for the long comprehensible documentation, it's great!

**Comment by MikeMajara** [comment]
Does the documentation say at any point that commands and messages are processed in the order they are written in the code?

**Comment by samson-sham** [comment]
In reply to @IMalyugin 
> **How are Scenes attached to telegraf bot?**
> The foundational component of Scenes is a `Stage`,
> 
> ```
> import { Stage } from 'telegraf';
> const stage = new Stage([scene1, scene2, scene3, ...]);
> bot.use(session()); // to  be precise, session is not a must have for Scenes to work, but it sure is lonely without one
> bot.use(stage.middleware());
> ```

`bot.use(session())` is compulsory for stage.middleware to inject its context into the TelegrafContext, if without initiate a session, the context passed has no scene, thus you cannot enter any scene.

**Comment by Romans96** [comment]
Is it possible to leave a WizardScene, during a step? For example by a command or a button attached to the current step reply?

**Comment by foodornt** [comment]
Anyway to use typescript with the wizard scene?

**Comment by dave-dm** [comment]
official documentation for scenes is needed asap. there's like 1 example and it explains very little.

**Comment by Omkaragrawal** [comment]
> import { composeWizardScene } from '../../utils/sceneFactory';
> 
> export const createContactDataScene = composeWizardScene(
>   (ctx) => {
>     ctx.reply('Please enter your credentials');
>     return ctx.wizard.next();
>   },
>   (ctx, done) => {
>     ctx.wizard.state.credentials = ctx.message.text;
>     return done();
>   },
> );

Hey @IMalyugin ,
To understand clearly, I am asking that is this approach of composeWizardScene and createWizardScene similar to that of redux? with a difference of instead of one function (scene) as a parameter, here we pass multiple functions (scenes')?

**Comment by Omkaragrawal** [comment]
> function createWizardScene(sceneType, nextScene) {
>     return new WizardScene(
>       sceneType,
>       ...advancedSteps.map((stepFn) => async (ctx, next) => {
>         /** ignore user action if it is neither message, nor callbackQuery */
>         if (!ctx.message && !ctx.callbackQuery) return undefined;
>         return stepFn(ctx, () => unwrapCallback(ctx, nextScene), next);
>       }),
>     );
>   }
> );

Hey @IMalyugin ,
To understand clearly, I am asking that is this approach of composeWizardScene and createWizardScene similar to that of redux? with a difference of instead of one function (scene) as a parameter, here we pass multiple functions (scenes')?



> export const createContactDataScene = composeWizardScene(
>   (ctx) => {
>     ctx.reply('Please enter your credentials');
>     return ctx.wizard.next();
>   },
>   (ctx, done) => {
>     ctx.wizard.state.credentials = ctx.message.text;
>     return done();
>   },
> );

**Comment by wojpawlik** [comment]
#1194 added examples for [scenes](https://github.com/telegraf/telegraf/tree/develop/docs/examples/scenes) and [wizards](https://github.com/telegraf/telegraf/tree/develop/docs/examples/wizards).

I think I'll treat https://github.com/telegraf/telegraf/issues/705#issuecomment-549056045 as our scenes docs for now and link to it from readme, so it'd be nice if you added type annotations to it after #1266 @KnorpelSenf.
